cluster-autoscaler/clusterstate/api/utils.go

```diff
@@ -90,3 +99,65 @@ func (status ClusterAutoscalerStatus) GetReadableString() string {
 	}
 	return buffer.String()
 }
+
+// LogItem is a single entry in log managed by LogCollector.
+type LogItem struct {
+	// Log is the logged message body.
+	Log string
+	// Timestamp when the Log was created.
+	Timestamp time.Time
+}
+
+// LogCollector keeps recent log events. It is automatically truncated on each access based on predefined set of conditions.
+type LogCollector struct {
+	sync.Mutex
+	// MaxItems is the maximum size of the log. Adding further messages will cause the oldest ones to be dropped.
+	MaxItems int
+	// ItemLifetime is the maximum time an item can be stored in log.
+	ItemLifetime time.Duration
+	// Store keeps LogItems.
+	Store []LogItem
+}
+
+// NewLogCollector creates new LogCollector.
+func NewLogCollector() *LogCollector {
+	return &LogCollector{
+		MaxItems:     DefaultLogCollectorMaxItems,
+		ItemLifetime: DefaultLogCollectorItemLifetime,
+		Store:        make([]LogItem, 0)}
+}
+
+// To be executed under a lock.
+func (lc *LogCollector) compact(now time.Time) {
+	firstIndex := 0
+	if len(lc.Store) > lc.MaxItems {
+		firstIndex = len(lc.Store) - lc.MaxItems
+	}
+	for ; firstIndex < len(lc.Store); firstIndex++ {
+		lifetimeEnd := lc.Store[firstIndex].Timestamp.Add(lc.ItemLifetime)
```

Comment:
what about:
threshold := now.Add(-lc.ItemLifetime)
...
if lc.Store[firstIndex].After(threshold) { break }

In the current form it is a bit tricky to understand.